modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/BucketedChildReferences.java

```diff
@@ -0,0 +1,375 @@
+import org.modeshape.jcr.value.Name;
+import org.modeshape.jcr.value.NameFactory;
+import org.modeshape.jcr.value.NamespaceRegistry;
+import org.modeshape.jcr.value.ValueFactory;
+
+/**
+ * A child references implementation which uses separate buckets for groups of children, based on the SHA1 of the name of each 
+ * child and the an order of magnitude constant (named 'prb') stored in the parent document. 
+ * See <a href='https://issues.jboss.org/browse/MODE-2109'>MODE-2109</a>
+ * 
+ * @author Horia Chiorean (hchiorea@redhat.com)
+ */
+class BucketedChildReferences extends AbstractChildReferences {
+    
+    private final long size;
+    private final int prb;
+    private final String parentKey;
+    private final DocumentTranslator translator;
+    private final LinkedHashMap<BucketId, Bucket> rangeBucketsById;
+    private final Set<BucketId> bucketIds; 
+    
+    protected BucketedChildReferences( Document parent, DocumentTranslator translator ) {
+        // the power of 16 which indicates how many buckets
+        this.prb = parent.getInteger(DocumentConstants.PRB);
+        this.size = parent.getLong(DocumentConstants.SIZE);
+        this.parentKey = translator.getKey(parent);
+        this.translator = translator;
+        List<?> bucketsArray = parent.getArray(DocumentConstants.BUCKETS);
+        this.bucketIds = new HashSet<>(bucketsArray.size());
+        for (Object bucketId : bucketsArray) {
+            this.bucketIds.add(new BucketId(bucketId.toString()));
+        }
+        this.rangeBucketsById = new LinkedHashMap<>(bucketIds.size());
+    }
+
+    @Override
+    public long size() {
+        return size;
+    }
+    
+    @Override
+    public int getChildCount( Name name ) {
+        Bucket bucket = bucketFor(name);
+        return bucket != null ? bucket.childCount(name) : 0;
+    }
+
+    @Override
+    public ChildReference getChild( Name name, int snsIndex, Context context ) {
+        if (snsIndex != 1) {
+            // we don't support SNS
+            return null;
+        }
+        Bucket bucket = bucketFor(name);
+        if (context == null || context.changes() == null) {
+            // there are no additional changes
+            return bucket != null ? bucket.childReferenceFor(name) : null;
+        }
+        // there are changes
+        Changes changes = context.changes();
+        ChildReference ref = bucket != null ? bucket.childReferenceFor(name) : null;
+
+        if (ref == null) {
+            // we don't have a reference in our data, look at the changes
+            Iterator<ChildInsertions> iterator = changes.insertions(name);
+            if (iterator.hasNext()) {
+                ChildInsertions inserted = iterator.next();
+                Iterator<ChildReference> iter = inserted.inserted().iterator();
+                // we don't support SNS, renames or reorderings so we'll always return the fist insertion
+                return iter.hasNext() ? iter.next() : null;
```

Comment:
`changes.insertions(name)` will only return insertions with that certain name. If those exist, they must be it because SNS are not supported.